### PR TITLE
Demo_User CRUD API

### PR DESCRIPTION
**Pull request by AyushSharma830**
Get all Users, get a particular user and create a new user APIs

### CODE REVIEW

**Comment by AyushSharma830** [comment]
Completed CRUD for Demo_User

File: 4. User API/userAPI/src/main/kotlin/Main.kt [hunk 1]
```diff
@@ -0,0 +1,19 @@
+
```
Comment (virendraOfficial) [inline 4. User API/userAPI/src/main/kotlin/Main.kt:1]: remove extra files from being pushed onto git, i.e. add them into gitignore

File: 4. User API/userAPI/src/main/kotlin/User.kt [hunk 2]
```diff
@@ -0,0 +1 @@
+data class User(val id : Int?, val name : String?, val age : Int?)
```
Comment (virendraOfficial) [inline 4. User API/userAPI/src/main/kotlin/User.kt]: make model with proper constructor and toString()

File: 4. User API/userAPI/src/main/kotlin/UserRepository.kt [hunk 3]
```diff
@@ -0,0 +1,27 @@
+class UserRepository {
```
Comment (virendraOfficial) [inline 4. User API/userAPI/src/main/kotlin/UserRepository.kt]: every method should have at-most 80 lines of code in it, if it is exceeding then create another method.
Every class must have at-most 500 lines of code, else make another class/service.
download sonarlint and install it in your ide to get indications to write cleaner and readable code and reduce cognitive complexity

File: 4. User API/userAPI/src/main/kotlin/UserRepository.kt [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    fun getUserById(id : Int ) : User? {
+        return users.firstOrNull{ it.id == id }
+    }
+
+    fun addUser(user : User ) : User? {
+        users.add(user)
+        return users.firstOrNull{ it.id == user.id }
+    }
+
+    fun updateUser(existingUser: User, newUser : User) : User? {
```
Comment (virendraOfficial) [inline 4. User API/userAPI/src/main/kotlin/UserRepository.kt]: use indexOfFirst{} lambda function.

File: 4. User API/userAPI/src/main/kotlin/UserResource.kt [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        val user : User? = userService.getUserById(id)
+        return if(user == null)    {
+            ResponseUtil.toResponse(status = 400, data = null, error ="Bad Request. No Such User Found.").toString()
+        } else{
+            ResponseUtil.toResponse(status = 200, data = user, error = null).toString()
+        }
+    }
+
+    @POST                       //CREATE New User
+    @Path("/add")
```
Comment (virendraOfficial) [inline 4. User API/userAPI/src/main/kotlin/UserResource.kt]: /user

File: 4. User API/userAPI/src/main/kotlin/UserResource.kt [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+            if(user != null) {
+                ResponseUtil.toResponse(status = 200, data = user, error = null).toString()
+            }else{
+                ResponseUtil.toResponse(status = 500, data = null, error = "Internal Server Error").toString()
+            }
+        }
+    }
+
+    @PUT                //UPDATE A Particular User
+    @Path("/update/{id}")
```
Comment (virendraOfficial) [inline 4. User API/userAPI/src/main/kotlin/UserResource.kt]: user

File: 4. User API/userAPI/src/main/kotlin/UserRepository.kt [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+    fun getAllUsers() : List<User> {
+        return users
+    }
+
+    fun getUserById(id : Int ) : User? {
+        return users.firstOrNull{ it.id == id }
+    }
+
+    fun addUser(user : User ) : User? {
```
Comment (virendraOfficial) [inline 4. User API/userAPI/src/main/kotlin/UserRepository.kt]: generate unique userId (uuid) to identify uniquely a records in the system.

File: 4. User API/userAPI/src/main/kotlin/UserService.kt [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+        val users  = userRepository.getAllUsers()
+        return users
+    }
+
+    fun getUserById(id : Int) : User?{
+        return userRepository.getUserById(id)
+    }
+
+    fun addUser(userData : String) : User? {
+        val user  = Gson().fromJson(userData,User::class.java)
```
Comment (virendraOfficial) [inline 4. User API/userAPI/src/main/kotlin/UserService.kt]: apply validations of user payload, also apply user system uniqueness

File: 4. User API/userAPI/src/main/kotlin/UserService.kt [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+        val addedUser  = userRepository.addUser(user)
+        return addedUser
+    }
+
+    fun updateUser(existingUser: User, userData : String) : User? {
```
Comment (virendraOfficial) [inline 4. User API/userAPI/src/main/kotlin/UserService.kt]: apply validations , check for uniqueness.

File: 4. User API/userAPI/src/main/kotlin/utility/ResponseUtil.kt [hunk 10]
```diff
@@ -0,0 +1,7 @@
+package utility
+
+object ResponseUtil {
```
Comment (virendraOfficial) [inline 4. User API/userAPI/src/main/kotlin/utility/ResponseUtil.kt:3]: you can make 2 methods, one to denote success response and another to denote failure response.

File: 4. User API/userAPI/src/main/kotlin/UserResource.kt [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+                }else{
+                    ResponseUtil.toResponse(status = 500, data = null, error = "Internal Server Error.").toString()
+                }
+            }
+        }else{
+            ResponseUtil.toResponse(status = 400, data = null, error ="Bad Request. No Such User Found.").toString()
+        }
+    }
+
+    @DELETE                         //DELETE A Particular User
```
Comment (virendraOfficial) [inline 4. User API/userAPI/src/main/kotlin/UserResource.kt]: if you want to add comments then can add method comment.
Also, read again about path and query param and the limitations of each.

**Pull request merged**